### ISSUE TITLE
[Question] Algorithm2 or can we neglect 1 and 4 steps ? 

### ISSUE THREAD

**Issue by vfdev-5** [ispgroupucl/layer-rotation-tools#1]
Hi,

According to the paper, snippet Algorithm 2:
> We’ve noticed that in practice, operations 1 and 4 are not strictly necessary for controlling layer rotation rates.

Does it make sense a minimal layca update as only 2, 3 steps:
- rotation-based normalization
- update

without 1 and 4 steps (as the code contains all steps) ?

Thanks

PS.:

In the snippets **Algorithm 1** and **Algorithm 2** the last operation (4 - project weights back on sphere) updates the final weights with weights at `t=0`, but in the code `initial_norm` correspond to weights at `t`:

https://github.com/ispgroupucl/layer-rotation-tools/blob/32d70f3f767f4ed8f2eb83aa5b6bc40fbdc82a0f/keras/layer_rotation_control.py#L32-L43

is it a typo ?

**Comment by Simoncarbo** [comment]
Hi!

I would only skip step 1, as step 4 can reduce numerical instabilities (the weights' norms tend to increase at each training step). This means commenting the following line:
`step = step - (K.sum(step * p))* p / initial_norm**2`
I'm not sure how much exactly you'll gain in efficiency by removing this.

Answer to your PS: It's not a typo. The norms of the weights at time `t` are the same as at time `t=0` in our case (since the norm is kept unchanged at each training step).

I hope this helps!

**Comment by vfdev-5** [comment]
@Simoncarbo thanks !

I close the issue as answered
